builtin/providers/librato/resource_librato_metric.go

```diff
@@ -0,0 +1,432 @@
+	}
+
+	if a, ok := d.GetOk("attributes"); ok {
+
+		attributeData := a.([]interface{})
+		if len(attributeData) > 1 {
+			return fmt.Errorf("Only one set of attributes per metric is supported")
+		}
+
+		if len(attributeData) == 1 && attributeData[0] == nil {
+			return fmt.Errorf("No attributes found in attributes block")
+		}
+
+		attributeDataMap := attributeData[0].(map[string]interface{})
+		attributes := new(librato.MetricAttributes)
+
+		if v, ok := attributeDataMap["color"].(string); ok && v != "" {
+			attributes.Color = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_max"].(string); ok && v != "" {
+			attributes.DisplayMax = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_min"].(string); ok && v != "" {
+			attributes.DisplayMin = librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_units_long"].(string); ok && v != "" {
+			attributes.DisplayUnitsLong = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_units_short"].(string); ok && v != "" {
+			attributes.DisplayUnitsShort = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["created_by_ua"].(string); ok && v != "" {
+			attributes.CreatedByUA = *librato.String(v)
+		}
+		if v, ok := attributeDataMap["display_stacked"].(bool); ok {
+			attributes.DisplayStacked = *librato.Bool(v)
+		}
+		if v, ok := attributeDataMap["gap_detection"].(bool); ok {
+			attributes.GapDetection = *librato.Bool(v)
+		}
+		if v, ok := attributeDataMap["aggregate"].(bool); ok {
+			attributes.Aggregate = *librato.Bool(v)
+		}
+
+		metric.Attributes = attributes
+	}
+
+	_, err := client.Metrics.Edit(metric)
+	if err != nil {
+		log.Printf("[INFO] ERROR creating Metric: %s", err)
+		return fmt.Errorf("Error creating Librato metric: %s", err)
+	}
+
+	resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, _, err := client.Metrics.Get(*metric.Name)
+		if err != nil {
+			if errResp, ok := err.(*librato.ErrorResponse); ok && errResp.Response.StatusCode == 404 {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	return metricReadResult(d, metric)
+}
+
+func resourceLibratoMetricRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*librato.Client)
+
+	id := d.Id()
+
+	log.Printf("[INFO] Reading Librato Metric: %s", id)
+	metric, _, err := client.Metrics.Get(id)
+	if err != nil {
+		if errResp, ok := err.(*librato.ErrorResponse); ok && errResp.Response.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Librato Metric %s: %s", id, err)
+	}
+
+	log.Printf("[INFO] Read Librato Metric: %s", structToString(metric))
+
+	return metricReadResult(d, metric)
+}
+
+func resourceLibratoMetricUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*librato.Client)
+
+	metricID := d.Id()
+
+	// Just to have whole object for comparison before/after update
+	fullMetric, _, err := client.Metrics.Get(metricID)
```

Comment:
I must admit, I was using the `resourceLibratoAlert` as a guide when I wrote this method. I'm going to review how other providers handle update, rewrite this method and simplify it as much as possible. This is a mess.